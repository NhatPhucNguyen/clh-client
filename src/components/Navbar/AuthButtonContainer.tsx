import { useNavigate } from "react-router-dom";
import styled from "styled-components";
import { useAuthContext } from "../../context/AuthenticationContext";

const ButtonContainer = styled.div`
    margin: auto auto 0 auto;
    width: clamp(5rem,8rem,15rem);
`;

const Button = styled.button`
    width: 100%;
    border-radius: 10px;
    border: none;
    padding: 0.2rem 0;
    background-color: #608ea1;
    color: #ffffff;
    font-weight: 600;
    &:hover {
        cursor: pointer;
        background-color: #29799b;
    }
`;
const AuthButtonContainer = () => {
    const navigate = useNavigate();
    const { isAuthenticated } = useAuthContext();
    return (
        <ButtonContainer>
            {isAuthenticated ? (
                <Button
                    onClick={() => {
                        navigate("/");
                    }}
                >
                    Logout
                </Button>
            ) : (
                <Button
                    onClick={() => {
                        navigate("/auth/login");
                    }}
                >
                    Login
                </Button>
            )}
        </ButtonContainer>
    );
};

export default AuthButtonContainer;
