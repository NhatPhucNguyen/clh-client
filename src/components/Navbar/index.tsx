import { faBars } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useState } from "react";
import { NavLink as DomNavLink, useNavigate } from "react-router-dom";
import styled, { keyframes } from "styled-components";
import { devices } from "../../config/devices";
import AuthContextProvider from "../../context/AuthenticationContext";
import AuthButtonContainer from "./AuthButtonContainer";
const fadeLeftIn = keyframes`
  from{
    opacity: 0;
    transform: translateX(100px);
  }
  to{
    opacity: 1;
    transform: translateX(0);
  }
`;
const Nav = styled.nav`
    display: block;
    padding: 0 5%;
    background-color: rgb(157, 196, 212);
    box-shadow: rgba(96, 96, 96, 0.04) 0px 2px 4px;
    @media screen and (${devices.tablets}) {
        padding: 0 10%;
    }
`;
const NavHeading = styled.div`
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    width: 100%;
    height: 20%;
`;
const SubNav = styled.div<{ $show: boolean }>`
    display: ${(props) => (props.$show ? "flex" : "none")};
    flex-direction: column;
    position: fixed;
    right: 0;
    height: 85%;
    background-color: #ffffff;
    width: clamp(12rem, 15rem, 70%);
    border-top-left-radius: 10px;
    border-bottom-left-radius: 10px;
    overflow-y: auto;
    animation: ${fadeLeftIn} 0.3s linear;
    padding: 1rem;
    z-index: 1;
    @media screen and (${devices.tablets}) {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        align-items: center;
        margin-top: 1rem;
        position: inherit;
        width: 100%;
        background-color: inherit;
        animation: none;
        border-radius: none;
        padding: 0;
        padding-bottom: 1rem;
    }
`;
const NavLinksContainer = styled.div`
    display: flex;
    flex-direction: column;
    gap: 1rem;
    width: 100%;
    @media screen and (${devices.tablets}) {
        display: flex;
        flex-direction: row;
        gap: 5%;
    }
`;
const LogoContainer = styled.div`
    display: flex;
    align-items: center;
    min-height: 15%;
    max-height: 20%;
`;
const Logo = styled.img`
    width: 15%;
    max-width: 800px;
    &:hover {
        cursor: pointer;
    }
`;
const CompanyName = styled.div`
    font-size: inherit;
    font-weight: bold;
    color: #31718c;
    letter-spacing: 1px;
`;
const Hamburger = styled.button<{ $isShowNav: boolean }>`
    background-color: #6d94a4;
    padding: 7px 10px;
    height: max-content;
    border-radius: 50%;
    transition: all 0.2s linear;
    border: none;
    transform: rotate(${(props) => (props.$isShowNav ? "90deg" : "0")});
    &:hover {
        cursor: pointer;
        background-color: #31718c;
    }
    @media screen and (${devices.tablets}) {
        display: none;
    }
`;
const NavLink = styled(DomNavLink)`
    text-decoration: none;
    border-bottom: 1px solid gray;
    font-weight: 500;
    color: #000000;
    &.active {
        color: #0087c2;
    }
    &:hover {
        cursor: pointer;
        color: #0087c2;
    }
    @media screen and (${devices.tablets}) {
        border-bottom: none;
    }
`;
const Navbar = () => {
    //state for small devices
    const [showNav, setShowNav] = useState(false);
    const navigate = useNavigate();
    return (
        <Nav>
            <NavHeading>
                <LogoContainer>
                    <Logo
                        src="icon.png"
                        role="logo"
                        onClick={() => {
                            navigate("/");
                        }}
                    />
                    <CompanyName>CodeLitHub</CompanyName>
                </LogoContainer>
                <Hamburger
                    onClick={() => {
                        setShowNav(!showNav);
                    }}
                    $isShowNav={showNav}
                >
                    <FontAwesomeIcon icon={faBars} size="lg" color="#ffffff" />
                </Hamburger>
            </NavHeading>
            <SubNav $show={showNav}>
                <NavLinksContainer>
                    <NavLink role="test" to={"/"} reloadDocument>
                        Home
                    </NavLink>
                    <NavLink to={"/books"}>Books</NavLink>
                    <NavLink to={"/subjects"}>Subjects</NavLink>
                    <NavLink to={"/authors"}>Authors</NavLink>
                    <NavLink to={"/bestsellers"}>Bestsellers</NavLink>
                </NavLinksContainer>
                <AuthContextProvider>
                    <AuthButtonContainer />
                </AuthContextProvider>
            </SubNav>
        </Nav>
    );
};

export default Navbar;
