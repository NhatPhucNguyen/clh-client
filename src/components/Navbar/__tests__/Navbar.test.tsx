import { render } from "@testing-library/react";
import { describe, test } from "vitest";
import { MemoryRouter } from "react-router-dom";
import Navbar from "..";

describe("(Navbar) Component", () => {
    test("Display all required nav link and logo", () => {
        const navBar = render(
            <MemoryRouter>
                <Navbar />
            </MemoryRouter>
        );
        expect(navBar).toMatchSnapshot();
    });
});
