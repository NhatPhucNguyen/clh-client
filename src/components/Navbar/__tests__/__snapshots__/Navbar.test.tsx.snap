// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`(Navbar) Component > Display all required nav link and logo 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="sc-dmyCSP eBGlef"
      >
        <div
          class="sc-hLQSwg gweSNH"
        >
          <div
            class="sc-gLLuof hhZzCU"
          >
            <img
              class="sc-iBdnpw fELnyv"
              role="logo"
              src="icon.png"
            />
            <div
              class="sc-fsYfdN jyKnsk"
            >
              CodeLitHub
            </div>
          </div>
          <button
            class="sc-qZrbh gppiFQ"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-bars fa-lg "
              color="#ffffff"
              data-icon="bars"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 96C0 78.3 14.3 64 32 64H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32H32c-17.7 0-32-14.3-32-32s14.3-32 32-32H416c17.7 0 32 14.3 32 32z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
        <div
          class="sc-eDLKkx kRuABz"
        >
          <div
            class="sc-jTQCzO gjDVkm"
          >
            <a
              aria-current="page"
              class="sc-jsEeTM cTCtiY active"
              href="/"
              role="test"
            >
              Home
            </a>
            <a
              class="sc-jsEeTM cTCtiY"
              href="/books"
            >
              Books
            </a>
            <a
              class="sc-jsEeTM cTCtiY"
              href="/subjects"
            >
              Subjects
            </a>
            <a
              class="sc-jsEeTM cTCtiY"
              href="/collections"
            >
              Collections
            </a>
            <a
              class="sc-jsEeTM cTCtiY"
              href="/bestsellers"
            >
              Bestsellers
            </a>
          </div>
          <div
            class="sc-beySPh fwJDGo"
          >
            <button
              class="sc-guDLey kbNKjP"
            >
              Login
            </button>
          </div>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="sc-dmyCSP eBGlef"
    >
      <div
        class="sc-hLQSwg gweSNH"
      >
        <div
          class="sc-gLLuof hhZzCU"
        >
          <img
            class="sc-iBdnpw fELnyv"
            role="logo"
            src="icon.png"
          />
          <div
            class="sc-fsYfdN jyKnsk"
          >
            CodeLitHub
          </div>
        </div>
        <button
          class="sc-qZrbh gppiFQ"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-bars fa-lg "
            color="#ffffff"
            data-icon="bars"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 96C0 78.3 14.3 64 32 64H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32H32c-17.7 0-32-14.3-32-32s14.3-32 32-32H416c17.7 0 32 14.3 32 32z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <div
        class="sc-eDLKkx kRuABz"
      >
        <div
          class="sc-jTQCzO gjDVkm"
        >
          <a
            aria-current="page"
            class="sc-jsEeTM cTCtiY active"
            href="/"
            role="test"
          >
            Home
          </a>
          <a
            class="sc-jsEeTM cTCtiY"
            href="/books"
          >
            Books
          </a>
          <a
            class="sc-jsEeTM cTCtiY"
            href="/subjects"
          >
            Subjects
          </a>
          <a
            class="sc-jsEeTM cTCtiY"
            href="/collections"
          >
            Collections
          </a>
          <a
            class="sc-jsEeTM cTCtiY"
            href="/bestsellers"
          >
            Bestsellers
          </a>
        </div>
        <div
          class="sc-beySPh fwJDGo"
        >
          <button
            class="sc-guDLey kbNKjP"
          >
            Login
          </button>
        </div>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
